examples/preproc3-NVT_E2E.ipynb

```diff
@@ -0,0 +1,890 @@
+    "train_set = (read_csv(DATA_BUCKET_FOLDER+'clicks_train.csv')\n",
+    "             .merge(read_csv(DATA_BUCKET_FOLDER + 'events.csv'), on=\"display_id\", how=\"left\", suffixes=('', '_event'))\n",
+    "             .merge(read_csv(DATA_BUCKET_FOLDER+'promoted_content.csv'), on=\"ad_id\", how=\"left\", suffixes=('', '_promo'))\n",
+    "             .merge(documents_meta, on=\"document_id\", how=\"left\")\n",
+    "             .merge(documents_meta, left_on=\"document_id_promo\", right_on=\"document_id\", how=\"left\", suffixes=('', \"_promo\"))\n",
+    "            )\n",
+    "train_set.to_parquet(\"train_pdf.parquet\", compression=None)\n",
+    "train_set = None"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "documents_categories_cudf = cudf.read_csv(\n",
+    "    DATA_BUCKET_FOLDER + 'documents_categories.csv',\n",
+    "    names=['document_id', 'category_id', 'confidence_level'],\n",
+    "    dtype={'document_id': 'int64','category_id': 'int64','confidence_level': 'float32'},\n",
+    "    skiprows=1)\n",
+    "\n",
+    "documents_topics_cudf = cudf.read_csv(\n",
+    "    DATA_BUCKET_FOLDER + 'documents_topics.csv',\n",
+    "    names=['document_id', 'topic_id', 'confidence_level'],\n",
+    "    dtype={'document_id': 'int64','topic_id': 'int64','confidence_level': 'float32'},\n",
+    "    skiprows=1)\n",
+    "\n",
+    "documents_entities_cudf = cudf.read_csv(\n",
+    "    DATA_BUCKET_FOLDER + 'documents_entities.csv',\n",
+    "\n",
+    "    names=['document_id', 'entity_id', 'confidence_level'],\n",
+    "    dtype={'document_id': 'int64','entity_id': 'str','confidence_level': 'float32'},\n",
+    "    skiprows=1)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 30,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# read in document categories/topics/entities as cupy sparse matrices\n",
+    "def df_to_coo(df, row=\"document_id\", col=None, data=\"confidence_level\"):\n",
+    "    return cupy.sparse.coo_matrix((df[data].values, (df[row].values, df[col].values)))\n",
+    "\n",
+    "categories = df_to_coo(documents_categories_cudf, col=\"category_id\")\n",
+    "topics = df_to_coo(documents_topics_cudf, col=\"topic_id\")\n",
+    "documents_entities_cudf['entity_id'] = documents_entities_cudf['entity_id'].astype(\"category\").cat.codes\n",
+    "entities = df_to_coo(documents_entities_cudf, col=\"entity_id\")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "HASH_BUCKET_SIZES = {\n",
+    "    'document_id': 300000,\n",
+    "    'ad_id': 250000,\n",
+    "    'document_id_promo': 100000,\n",
+    "    'source_id_promo': 4000,\n",
+    "    'source_id': 4000,\n",
+    "    'advertiser_id': 2500,\n",
+    "    'publisher_id_promo': 1000,\n",
+    "    'publisher_id': 1000,\n",
+    "}"
```

Comment:
Do we want to use hashing instead of Categorify here?

Reply:
Yes we use HashBucket with these features.